pkg/token/filecache.go

```diff
@@ -0,0 +1,277 @@
+	clusterID, profile, roleARN string                   // parameters used to create Provider
+	cachedToken                 cacheToken               // the cached token, if it exists
+}
+
+// NewFileCacheProvider creates a new Provider implementation that wraps a provided Credentials,
+// and works with an on disk cache to speed up token usage when the cached copy is not expired.
+// If there are any problems accessing or initializing the cache, an error will be returned, and
+// callers should just use the existing credentials provider.
+func NewFileCacheProvider(clusterID, profile, roleARN string, creds *credentials.Credentials) (FileCacheProvider, error) {
+	filename := CacheFilename()
+	cachedToken := cacheToken{}
+	if info, err := f.Stat(filename); !os.IsNotExist(err) {
+		if info.Mode() & 0077 != 0 {
+			// cache file has secret credentials and should only be accessible to the user, refuse to use it.
+			return FileCacheProvider{}, fmt.Errorf("cache file %s is not private", filename)
+		}
+
+		// do file locking on cache to prevent inconsistent reads
+		lock := newFlock(filename)
+		defer lock.Unlock()
+		// wait up to a second for the file to lock
+		ctx, cancel := context.WithTimeout(context.TODO(), time.Second)
+		defer cancel()
+		ok, err := lock.TryRLockContext(ctx, 250 * time.Millisecond) // try to lock every 1/4 second
+		if !ok {
+			// unable to lock the cache, something is wrong, refuse to use it.
+			return FileCacheProvider{}, fmt.Errorf("unable to read lock file %s: %v", filename, err)
+		}
+
+		cache, err := readCacheWhileLocked(filename)
+		if err != nil {
+			// can't read or parse cache, refuse to use it.
+			return FileCacheProvider{}, err
+		}
+
+		if _, ok := cache.ClusterMap[clusterID]; ok {
+			if _, ok := cache.ClusterMap[clusterID][profile]; ok {
+				// we at least have this cluster and profile combo in the map, if no matching roleARN, map will
+				// return the zero-value for cacheToken, which expired a long time ago.
+				cachedToken = cache.ClusterMap[clusterID][profile][roleARN]
+			}
+		}
+	} else {
+		// cache file is missing.  maybe this is the very first run?  continue to use cache.
+		_, _ = fmt.Fprintf(os.Stderr, "Cache file %s does not exist.\n", filename)
+	}
+
+	return FileCacheProvider{
+		creds,
+		clusterID,
+		profile,
+		roleARN,
+		cachedToken,
+	}, nil
+}
+
+// Retrieve() implements the Provider interface, returning the cached token if is not expired,
+// otherwise fetching the token from the underlying Provider and caching the results on disk
+// with an expiration time.
+func (f *FileCacheProvider) Retrieve() (credentials.Value, error) {
+	if !f.cachedToken.IsExpired() {
+		// use the cached token
+		return f.cachedToken.Token, nil
+	} else {
+		_, _ = fmt.Fprintf(os.Stderr, "No cached token available.  Refreshing...\n")
+		// fetch the token from the underlying Provider
+		token, err := f.credentials.Get()
+		if err == nil {
```

Comment:
Can we reverse these `if` statements to remove this nesting?  Capture error and return at the top of the block.